enhancements/oc/mirroring-manifests.md

```diff
@@ -0,0 +1,172 @@
+This enhancement would also allow `oc apply`'s `--force-conflicts` and `--overwrite` options to allow users to manage conflicts with resources that already existed in the target cluster.
+If `oc` encountered an error applying a manifest, it could optionally attempt to apply additional manifests before exiting non-zero, but would not be required to attempt any additional manifests.
+
+### Pushing to disk
+
+As of 4.3:
+
+```console
+$ oc adm release mirror --to-dir=mirror quay.io/openshift-release-dev/ocp-release:4.3.0-x86_64
+...
+Success
+Update image:  openshift/release:4.3.0
+
+To upload local images to a registry, run:
+
+    oc image mirror --from-dir=mirror file://openshift/release:4.3.0* REGISTRY/REPOSITORY
+```
+
+creates an output filesystem like:
+
+```console
+$ tree mirror
+mirror
+└── v2
+    └── openshift
+        └── release
+            ├── blobs
+            │   ├── sha256:005337dc1f6870934cf5efaf443786149e6a71041a91566ce1e16e78bce511a9
+            ...
+            │   └── sha256:ffc132fefac522ec38b1b22ecd67a5173565efb3e8929e47ef7e0b4ee7920adf
+            └── manifests
+                ├── 4.3.0 -> mirror/v2/openshift/release/manifests/sha256:3a516480dfd68e0f87f702b4d7bdd6f6a0acfdac5cd2e9767b838ceede34d70d
+                ...
+                ├── 4.3.0-thanos -> mirror/v2/openshift/release/manifests/sha256:42ab3f59d3e769e88af89c30304b9cc77a76b66960747e2e1a6f8db097420858
+                ├── sha256:00ec58112784d340179e045bc70399ac1bd509ae43c4411e934ced23d016b2a1
+                ...
+                └── sha256:ffa87c992c5a2c51290e6d67cb609f29893dd147c31cff639ff6785ae7a1cfe2
+
+5 directories, 514 files
+```
+
+that can be applied to the local registry with the given `oc image mirror` command.
+
+With this enhancement, the output format would gain an additional directory `manifests` as a sibling to the current `v2` containing manifest files that can be applied to the local cluster with:
+
+```console
+$ oc apply -Rf mirror/manifests
+```
+
+The `manifests` directory may or may not contain additional subdirectories, so the recursive `-R` should be given for future-proofing.
+
+### User Stories
+
+#### Release-signature ConfigMaps
+
+FIXME
+
+#### In-cluster Cincinnati
+
+FIXME
+
+### Risks and Mitigations
+
+FIXME
+
+## Design Details
+
+### Test Plan
+
+FIXME
+
+### Graduation Criteria
+
+FIXME
+
+### Upgrade / Downgrade Strategy
+
+FIXME
+
+### Version Skew Strategy
+
+The proposal for [applying directly to the target cluster](#applying-directly-to-the-target-cluster) adds a new dependency on the cluster to which `oc` connects for Kubernetes activity.
+It's possible that users who have been using the previous `oc adm release mirror` implementation (which did not push manifests) would be unaware of the change and end up accidentally pushing manifests into the wrong cluster.
```

Comment:
That's probably a blocker right there.

Reply:
> That's probably a blocker right there.

Workaround would be to add a flag to opt into pushing manifests.  But I don't like "are you sure?" options.  Ideally folks would be `--dry-run` or whatever first (although that's unlikely in scripts) and really, you shouldn't be giving your `oc` creds it doesn't need.  So personally I'm fine with this as it stands, but if folks want a flag for `--and-push-the-manifests-too-please` or whatever they want to call it, let me know and I can pinch my nose and add that ;).

Reply:
`oc adm release mirror` should output of what artifacts it writing to the filesystem.  CVO logs should point out that it is getting the manifest information from the configmaps e.g. the signatures are coming from configmap and not from `mirror.openshift.com`.

Reply:
Doesn't 'oc adm release mirror' need a flag to indicate whether you're dealing with manifests for both options- 1) outputting only 2) outputting and applying to a target server directly. I thought pulling manifests would always be an option or are we always going to pull them and then optionally allow the user to use them or not. Either way, i don't see a option flag on oc as a big deal.

Reply:
> `oc adm release mirror` should output of what artifacts it writing to the filesystem.

We could.  But folks can just look at the filesystem, right?  I think the concern here is in the push-straight-into-a-cluster case, and if the user is pointing the `mirror` call at the wrong cluster by accident, they are probably not going to be satisfied when they look at their script's archived logs and see us saying "here's all the stuff we pushed into your cluster, surprise!".  We could log something like "we're about to push the following objects into `$CLUSTER`, press `y` to continue" and train folks to use a new `-i` / `--interactive` if they want a guard (like `mv`, `rm`, etc.), but there's no way to swing that that will be seamless for folks that call `oc adm release mirror ...` from scripts with a kubeconfig pointing at the wrong cluster, short of not pushing manifests at all unless they set a flag like `--and-push-the-manifests-too-please` to opt in to the new functionality.

> CVO logs should point out that it is getting the manifest information from the configmaps e.g. the signatures are coming from configmap and not from mirror.openshift.com.

Do we care?  I'd expect us to care about "we found a valid sig" and "we found no valid sigs after looking in $`PLACES`".  But this seems like a purely-CVO issue, orthogonal to this `oc` enhancement.

> Doesn't `oc adm release mirror` need a flag to indicate whether you're dealing with manifests for both options- 1) outputting only 2) outputting and applying to a target server directly.

Yes, but as of ac55c2dd7, I'm piggybacking on the existing `--to-dir` for distinguish.  And I don't know why you'd want to *both* dump manifests to disk and apply to a target cluster.  Can you give a use case?

Reply:
> why you'd want to both dump manifests to disk and apply to a target cluster

No, that's not what I meant. Won't the users of 'oc adm release mirror' always have an option of whether to pull manifests? In other words, the signature functionality we're proposing to add may not always be needed by all users and therefore there already needs to be a flag on the command to indicate whether you want to pull manifests. If not then we will be pulling manifests for all users of this command - which may be okay.

Reply:
> In other words, the signature functionality we're proposing to add may not always be needed by all users and therefore there already needs to be a flag on the command to indicate whether you want to pull manifests.

In the `--to-dir` case, there's no harm in gathering manifests.  Users can always pick which ones they want to push.

In the direct-push case, you can block pushing by not providing `oc` with a kubeconfig.  Or by omitting an `--and-push-the-manifests-too-please` option (or whatever we want to call that).  Do folks have preferences for that option's name, since there seems to be a desire for it?

Reply:
>  Do folks have preferences for that option's name, since there seems to be a desire for it?

Ok, added `--apply-manifests` with ac55c2d -> 8d55831.  Marking this thread as resolved, but feel free to reopen if folks feel like there are additional points from this discussion that still need to be addressed.